Support requests with non-'utf-8' charsets

theopak opened the pull request:
~~_Not ready to merge._~~ ~~EDIT: Ready, pending approval.~~ **UPDATE:** Ready.
#### TODO
- [x] "fix the silent failures in the Travis-CI build"
  - EDIT: thanks Ryan https://github.com/theopak/oauth_reverse_proxy/pull/4
- [x] skip `body_parser` middleware in case of empty body
- [x] add pre-mutator to stash original content type
- ~~add pre-mutator to convert content to UTF-8 so that `body_parser.urlencoded()` parses it~~
  - wontfix
- [x] add post-mutator to restore original content type
- [x] verify that non-`utf-8` charsets are no longer silently failed
- [x] verify that, if removed, non-`utf-8` charsets are always restored after `body_parser.urlencoded()` does its thing.
- verify that requests with `utf-8` query strings (e.g., emoji) are proxied correctly
  - [x] unit tests exist and pass

theopak committed:
Strip 'Content-Type' , and then add it again after body_parser does its thing.

theopak committed:
Code quality

theopak committed:
Combine pre- and post-urlencoded mutators into a single middleware.

theopak committed:
Add unit test to verify support of non-`utf-8` charset requests.

theopak commented on test/message_integrity_header_test.js:
```diff
@@ -100,6 +100,17 @@ require('./bootstrap_test.js');
+      it ("should support " + verb + " requests with non-'utf-8' charsets declared in the header", function(done) {
+        var intended_content_type_header = 'text/html; charset=ISO-8859-8';
+        var custom_headers = { headers: {'Content-Type': intended_content_type_header } };
+        request_sender.sendRequest(verb, 'http://localhost:8080/livecheck', custom_headers, 200, function(err, res, body) {
```
@ryanbreen What job_server endpoint would be appropriate to test against in this case?

ryanbreen replied:
I think POST /job would be good.  You want to make sure that a url-encoded post gets processed correctly.

theopak committed:
Fix build validity failure caused by https://github.com/theopak/oauth_reverse_proxy/commit/538607c70720461b98848307c5498e96cd2e8e4a

theopak committed:
Use a test endpoint that actually does work

theopak committed:
Require bash, because the calling shell may be something else (e.g. "dash" on the Travis-CI build).

theopak committed:
Remove bad test, add test that verifies form urlencoded PUT/POST requests in particular.

theopak committed:
Stash/unstash "Content-Encoding" and "Content-Type" headers. TODO: fix test case

theopak committed:
Expect these tests to fail due to https://github.com/request/request/commit/86571c794b75bb637b39eb7574e825e461647151

theopak commented:
It took me waaaay too long to figure this out, but request.js (since v2.64.1) automatically overrides form requests to have `Content-Type: "application/x-www-form-urlencoded"` when they are sent to the server. 

request.js will not send the kind of requests that this PR is looking to test. Therefore I'll have to take a different approach in forming the test.

theopak committed:
Override Request.js form header modification in request_sender.js which is used to form test requests

theopak committed:
Clean up

theopak committed:
Repeat request charset validation by explicitly each of the possible conditions.

theopak committed:
Don't stash charset === 'utf-8'.

theopak committed:
Remove Request.js customization

theopak committed:
Expand test cases

theopak commented:
@ryanbreen Could you take a look at the test case and make sure that it adequately captures the exact issue you told me about? My progress is stalled on this PR because I'm having trouble replicating the exact issue that you described to me in the chat last Wednesday.

theopak commented on test/message_integrity_formencoded_test.js:
```diff
@@ -1,10 +1,6 @@
 var path = require('path');
-
-var request_sender = require('./utils/request_sender.js');
-var validation_tools = require('./utils/validation_tools.js');
```
Removed because these are repeated in lines 8 and 9 (of the new file).

theopak commented on test/utils/request_custom.js:
```diff
@@ -0,0 +1,32 @@
+ * Comment out the parts of request.js that automatically set a form-urlencoded
+ * request to be `charset=utf-8` in this overriden defintion.
+ * Refer to https://github.com/request/request/commit/86571c794b75bb637b39eb7574e825e461647151
+ */
```
I stopped using this file and will remove it from the diff if it's definitely not needed.

Commit:
Restore java client.

theopak committed:
Failing test: enable header charset stash/unstash and content encoding conversion.

theopak committed:
Comment out previous commit, resulting in a test that causes body_parser.urlencoded() to throw an error due to unsupported charset.

theopak commented:
Inspecting `req.argument_pairs` during the unit tests, I notice that all the Java client's nonces are integers, and some of those are negative integers. This seems unusual to me. Two example values from failed requests:

``` bash
[2015-10-01T20:09:48.988Z] TRACE: oauth_reverse_proxy/54621 on lexmac1909.vistaprint.net: (service_name=jobsservice, module=lib/proxy/validators/oauth_signature_validator.js)
    req.argument_pairs:
    [ [ 'oauth_consumer_key', 'java-test-key' ],
      [ 'oauth_nonce', '1244941739633823073' ],
      [ 'oauth_signature_method', 'HMAC-SHA1' ],
      [ 'oauth_timestamp', '1443730188' ],
      [ 'oauth_version', '1.0' ],
      [ 'this', 'is' ],
      [ 'fun', 'right' ] ]
[2015-10-01T20:12:37.531Z] TRACE: oauth_reverse_proxy/54666 on lexmac1909.vistaprint.net: (service_name=jobsservice, module=lib/proxy/validators/oauth_signature_validator.js)
    req.argument_pairs:
    [ [ 'oauth_consumer_key', 'java-test-key' ],
      [ 'oauth_nonce', '-4825279163585566192' ],
      [ 'oauth_signature_method', 'HMAC-SHA1' ],
      [ 'oauth_timestamp', '1443730357' ],
      [ 'oauth_version', '1.0' ],
      [ 'this', 'is' ],
      [ 'fun', 'right' ] ]
```

"A nonce is a random string," according to http://oauth.net/core/1.0a/#nonce, which means that the other clients are following spec when they generate alphanumeric nonces as base64 encoded UTF-8 data. 

@RB Do you think this indicate a problem with the way that the client's requests are received?

ryanbreen commented:
I wouldn't think that would cause an issue.  An integer is still a random string.

Commit:
Fix Java test by forcing all formencoded posts to present a formencoded header without a charset.

Commit:
Restore all client tests.

theopak committed:
Code formatting

theopak committed:
Add stash/convert/unstash functions

theopak committed:
v1.5.10-4 multiple changes:

    - Remove `mutators/url_parser.js`.
    - Use `mutators/query_parser.js` for all URL query string parsing.
    - Add unit test for fancy characters (emoji) in URLs.
    - Add 'Usage' section to readme, recommending Unirest.
    - Install Perl test dependencies in Travis-CI build

theopak committed:
Add missing module_tag reference.

theopak committed:
Make Travis before_script faster

theopak committed:
fixup! Make Travis before_script faster

theopak committed:
Minor: Add missing require statement

theopak committed:
Fix Travis build error, Can't locate Net/OAuth.pm in @INC (@INC contains: /home/travis/perl5/lib/perl5/x86_64-linux-gnu-thread-multi /home/travis/perl5/lib/perl5 /etc/perl /usr/local/lib/perl/5.14.2 /usr/local/share/perl/5.14.2 /usr/lib/perl5 /usr/share/perl5 /usr/lib/perl/5.14 /usr/share/perl/5.14 /usr/local/lib/site_perl .) at client.pl line 5.

theopak committed:
Remove unused Request.js overload

theopak committed:
Fix 'should accept a properly signed GET|DELETE with {a funky query}|{no value}'.

theopak committed:
Minor: add logging, clarify wording of unit test

theopak committed:
Print env info in Travis build to help debug Perl dependencies in the future. Not spending any more time on this.

theopak committed:
Remove Perl client tests from Travis build.

theopak committed:
Client library tests pass as long as headers are removed for good, and never restored.

theopak commented:
@ryanbreen The client library tests pass, with your changes, as of https://github.com/Cimpress-MCP/oauth_reverse_proxy/commit/7675b31bb1c7a5f5d336f123472d0ad0d88791a7

However, your changes have a typo: https://github.com/theopak/oauth_reverse_proxy/pull/4#discussion_r41192734

The tests fail if I fix the typo

theopak committed:
Minor

Commit:
Also munge charsets on the jobserver test side.

theopak committed:
Log descriptive errors

theopak committed:
Merge commit '5a6f3af47e6c46a10641af9c505ec5eb904bb858' into bug/GH-15

* commit '5a6f3af47e6c46a10641af9c505ec5eb904bb858':
  Also munge charsets on the jobserver test side.

theopak committed:
TO-DONT do that

theopak committed:
Comment out some diffs

theopak committed:
fixup! Comment out some diffs

theopak committed:
Fix hardcoded value which made a test use GET twice instead of DELETE

theopak commented on test/oauth_proxy_outbound_test.js:
```diff
@@ -21,7 +21,7 @@ describe('oauth_proxy outbound message integrity: verbs', function() {
     ['GET', 'DELETE'].forEach(function(verb) {
```
The latter 4 test templates are the ones that still fail in the build, even though they pass on [master](https://travis-ci.org/Cimpress-MCP/oauth_reverse_proxy/builds/81434564).

``` bash
$ set OAUTH_REVERSE_PROXY_LOG_LEVEL="trace"; mocha --reporter spec -t 10000 ./test/oauth_proxy_outbound_test.js | bunyan

  64 passing (494ms)
  8 failing
```

@ryanbreen any idea why? I've tried rolling back my changes one by one locally, and looking through the diff in this PR, but I can't pinpoint the line(s) that cause these tests to fail.

ryanbreen replied:
All the failures are here:

```
8) should accept a properly signed DELETE over IPv6 with query via https
Error: Protocol "https:" not supported. Expected "http:".
    at new ClientRequest (_http_client.js:53:11)
    at Object.exports.request (http.js:31:10)
    at Object.exports.request (https.js:163:15)
    at Array.stream (/Users/wrb/tmp/oauth_reverse_proxy/node_modules/http-proxy/lib/http-proxy/passes/web-incoming.js:108:74)
    at ProxyServer.<anonymous> (/Users/wrb/tmp/oauth_reverse_proxy/node_modules/http-proxy/lib/http-proxy/index.js:80:21)
```

I'm not sure what might have changed between this version and master.  Looks like an https request is being sent with http.js.  Not sure why that would happen.

ryanbreen replied:
Looks like it might be a node > 0.11.0 discrepancy: https://github.com/SamDecrock/node-http-ntlm/issues/21

You might need to check the node-http-proxy version and see if there's an upgrade to address this, assuming the issue is from that package.

theopak replied:
Builds now thanks to your fix. Evidently the failures weren't directly related to using node > 0.11.0, but actually a change that happened when I upgraded dependencies and combined query_parser.js and url_parser.js without noticing all of the resulting differences.

Commit:
Make test less vulnerable to differences in output between versions.

theopak committed:
Merge pull request #6 from ryanbreen/bug/GH-15_rb

Make test less vulnerable to differences in output between versions.

Commit:
Set the proper agent for outbound requests.

theopak committed:
Merge pull request #7 from ryanbreen/bug/GH-15_rb

Set the proper agent for outbound requests.

theopak committed:
Cleanup

theopak committed:
Adds tests to validate that emoji are allowed in query strings

theopak committed:
Encode URL components

theopak committed:
Reorder test

ryanbreen commented:
LGTM!
